src/TypeScript/Rules/RuleDefinitionsProvider.cs

```diff
@@ -0,0 +1,80 @@
+﻿/*
+ * SonarLint for Visual Studio
+ * Copyright (C) 2016-2021 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Collections.Generic;
+using System.ComponentModel.Composition;
+using System.IO;
+using System.Linq;
+using System.Text;
+using Newtonsoft.Json;
+
+namespace SonarLint.VisualStudio.TypeScript.Rules
+{
+    internal interface ITypeScriptRuleDefinitionsProvider
+    {
+        /// <summary>
+        /// Returns the metadata descriptions for all TypeScript rules
+        /// </summary>
+        IEnumerable<RuleDefinition> GetDefinitions();
+    }
+
+    internal interface IJavaScriptRuleDefinitionsProvider
+    {
+        /// <summary>
+        /// Returns the metadata descriptions for all JavaScript rules
+        /// </summary>
+        IEnumerable<RuleDefinition> GetDefinitions();
+    }
+
+    [Export(typeof(ITypeScriptRuleDefinitionsProvider))]
+    [Export(typeof(IJavaScriptRuleDefinitionsProvider))]
+    [PartCreationPolicy(CreationPolicy.Shared)]
+    internal class RuleDefinitionsProvider : ITypeScriptRuleDefinitionsProvider, IJavaScriptRuleDefinitionsProvider
+    {
+        // Note: the file contains rules for both JavaScript and TypeScript rules
+        internal const string RuleDefinitionsFilePathContractName = "SonarLint.TypeScript.RuleDefinitionsFilePath";
+
+        private readonly IReadOnlyCollection<RuleDefinition> tsRules;
+        private readonly IReadOnlyCollection<RuleDefinition> jsRules;
+
+        private const string RepoPrefix_TypeScript = "typescript:";
+        private const string RepoPrefix_JavaScript = "javascript:";
+
+        [ImportingConstructor]
+        public RuleDefinitionsProvider([Import(RuleDefinitionsFilePathContractName)] string typeScriptMetadataFilePath)
+        {
+            var allRules = Load(typeScriptMetadataFilePath);
```

Comment:
Since this is loaded in the CTOR, if we fail to deserialize the json the user will see the extension load failure popup.

I think it's better to catch it and return empty rules. WDYT?

Reply:
Good point. Returning an empty list isn't going to give a great UX either, though. How will the user know that something is wrong? The other components will continue to work, the log will show that analysis is being triggered and no issues were found etc.
We'd have to make the caller check for an empty set of rules and warn the user somehow.

Something would have had to go badly wrong for us to ship a version with invalid json (both with the SonarJS code that generates the file, and our manual testing). We could also write more integration level tests to make sure the embedded file is valid, in the same way we do for the CFamily data files.

Issues could also occur at runtime on the user's machine e.g. due to file corruption/ a bad installation. I don't think we should try to carry on in that case, though.

For now, I'd rather let the extension fail if the file can't be deserialized.
I'll add a task to write a sanity check test for the embedded file.

How does that sound?

Reply:
Yep, we could do something like we do for CFamily. Maybe not just the rules, but just that the json is deserialized and the list is not empty for JS and TS. Sounds good.